File: refactoring/run_cluster_analysis.py
#!/usr/bin/env python3
"""
Pure-Python cluster analysis benchmark.

Reads prepared JSON from prepare_cluster_data.py and runs the full cluster
analysis loop (identical logic to src/core/datastore.py __process_single_attribute
+ __update_ALO_data), writing a compact result CSV.

Timing covers only the analysis loop, excluding file I/O.

Usage:
    python run_cluster_analysis.py cluster_data.json [--output results.csv]
"""
import argparse
import json
import sys
import time
from math import log, sqrt
from typing import Any, Dict, List, Optional, Set, Tuple

import scipy.stats


# ---------------------------------------------------------------------------
# Stat helpers (mirrors src/core/utils.py statistic())
# ---------------------------------------------------------------------------

def _mean(lst):
    return sum(lst) / len(lst)


def statistic(count_1, count_2, test, min_proteomes):
    pvalue = log2_mean = mean_c1 = mean_c2 = None
    c1 = [c for c in count_1 if c > 0]
    c2 = [c for c in count_2 if c > 0]
    if len(c1) < min_proteomes or len(c2) < min_proteomes:
        return None, None, None, None
    mean_c1 = _mean(c1)
    mean_c2 = _mean(c2)
    log2_mean = log(mean_c1 / mean_c2, 2)
    if len(set(c1)) == 1 and len(set(c2)) == 1 and set(c1) == set(c2):
        pvalue = 1.0
    elif test == "mannwhitneyu":
        try:
            pvalue = scipy.stats.mannwhitneyu(c1, c2, alternative="two-sided")[1]
        except ValueError:
            pvalue = 1.0
    elif test in ("welch", "ttest"):
        pvalue = scipy.stats.ttest_ind(c1, c2, equal_var=(test == "ttest"))[1]
        if pvalue != pvalue:
            pvalue = 1.0
    elif test == "ks":
        pvalue = scipy.stats.ks_2samp(c1, c2)[1]
        if pvalue != pvalue:
            pvalue = 1.0
    elif test == "kruskal":
        pvalue = scipy.stats.kruskal(c1, c2)[1]
        if pvalue != pvalue:
            pvalue = 1.0
    return pvalue, log2_mean, mean_c1, mean_c2


# ---------------------------------------------------------------------------
# Cardinality logic (mirrors src/core/logic.py get_ALO_cluster_cardinality)
# ---------------------------------------------------------------------------

def get_attribute_cluster_type(singleton, implicit_by_level):
    if singleton:
        return "singleton"
    return "shared" if len(implicit_by_level) > 1 else "specific"


def get_ALO_cluster_cardinality(alo_counts, fuzzy_count, fuzzy_fraction, fuzzy_range):
    if not alo_counts:
        return None
    non_zero = [c for c in alo_counts if c > 0]
    if not non_zero:
        return None
    if all(c == 1 for c in non_zero):
        return "true"
    if fuzzy_count in alo_counts or (
        len(non_zero) >= len(alo_counts) * fuzzy_fraction
        and all(c in fuzzy_range for c in non_zero)
    ):
        return "fuzzy"
    return None


# ---------------------------------------------------------------------------
# Length stats (mirrors ProteinCollection.get_protein_length_stats)
# ---------------------------------------------------------------------------

def get_protein_length_stats(protein_ids, protein_lengths, fastas_parsed):
    stats = {"sum": 0, "mean": 0.0, "median": 0, "sd": 0.0}
    if not protein_ids or not fastas_parsed:
        return stats
    lengths = [protein_lengths[pid] for pid in protein_ids if pid in protein_lengths and protein_lengths[pid] > 0]
    if not lengths:
        return stats
    stats["sum"] = sum(lengths)
    stats["mean"] = sum(lengths) / len(lengths)
    s = sorted(lengths)
    n = len(s)
    stats["median"] = (s[n // 2] + s[(n - 1) // 2]) / 2.0
    m = stats["mean"]
    stats["sd"] = sqrt(sum((x - m) ** 2 for x in lengths) / len(lengths)) if len(lengths) > 1 else 0.0
    return stats


# ---------------------------------------------------------------------------
# Main analysis loop
# ---------------------------------------------------------------------------

def run_analysis(data: dict) -> List[dict]:
    params = data["params"]
    test = params["test"]
    min_proteomes = params["min_proteomes"]
    fuzzy_count = params["fuzzy_count"]
    fuzzy_fraction = params["fuzzy_fraction"]
    fuzzy_range = set(params["fuzzy_range"])
    fastas_parsed = params["fastas_parsed"]

    alo_collection = data["alo_collection"]
    protein_lengths = data["protein_lengths"]
    attributes = list(alo_collection.keys())

    results = []

    for cluster in data["clusters"]:
        cluster_id = cluster["cluster_id"]
        singleton = cluster["singleton"]
        protein_ids_by_proteome_id = cluster["protein_ids_by_proteome_id"]

        for attribute in attributes:
            levels = alo_collection[attribute]

            # Pass 1: process each level — collect protein ids, counts, length stats
            protein_ids_by_level: Dict[str, List[str]] = {}
            protein_length_stats_by_level: Dict[str, dict] = {}
            explicit_count_by_proteome_by_level: Dict[str, Dict[str, int]] = {}
            implicit_by_level: Dict[str, dict] = {}  # level -> {proteome_id: [prot_ids]}

            for level, alo in levels.items():
                if alo is None:
                    continue
                proteomes_list = alo["proteomes_list"]

                pids_for_level = []
                count_by_proteome = {}
                implicit_for_level = {}

                for proteome_id in proteomes_list:
                    pids = protein_ids_by_proteome_id.get(proteome_id, [])
                    pids_for_level.extend(pids)
                    count_by_proteome[proteome_id] = len(pids)
                    if pids:
                        implicit_for_level[proteome_id] = pids

                protein_ids_by_level[level] = pids_for_level
                explicit_count_by_proteome_by_level[level] = count_by_proteome
                protein_length_stats_by_level[level] = get_protein_length_stats(
                    pids_for_level, protein_lengths, fastas_parsed
                )
                if implicit_for_level:
                    implicit_by_level[level] = implicit_for_level

            cluster_type = get_attribute_cluster_type(singleton, implicit_by_level)

            # Pass 2: per-level stats + result row
            for level, alo in levels.items():
                if alo is None:
                    continue
                proteome_count = alo["proteome_count"]

                alo_status = "present" if level in implicit_by_level else "absent"

                proteome_coverage = len(implicit_by_level.get(level, {})) / proteome_count

                cardinality = None
                mwu_pvalue = mwu_log2_mean = mean_alo = mean_non_alo = None

                if alo_status == "present" and cluster_type != "singleton":
                    alo_counts = list(explicit_count_by_proteome_by_level[level].values())
                    cardinality = get_ALO_cluster_cardinality(
                        alo_counts, fuzzy_count, fuzzy_fraction, fuzzy_range
                    )

                    if cluster_type == "shared":
                        non_alo_counts = [
                            c
                            for non_level, counts in explicit_count_by_proteome_by_level.items()
                            if non_level != level
                            for c in counts.values()
                        ]
                        mwu_pvalue, mwu_log2_mean, mean_alo, mean_non_alo = statistic(
                            alo_counts, non_alo_counts, test, min_proteomes
                        )

                results.append({
                    "cluster_id": cluster_id,
                    "attribute": attribute,
                    "level": level,
                    "cluster_type": cluster_type,
                    "alo_status": alo_status,
                    "proteome_coverage": proteome_coverage,
                    "cardinality": cardinality,
                    "mwu_pvalue": mwu_pvalue,
                    "mwu_log2_mean": mwu_log2_mean,
                    "mean_alo": mean_alo,
                    "mean_non_alo": mean_non_alo,
                    "protein_count": len(protein_ids_by_level.get(level, [])),
                    "length_sum": protein_length_stats_by_level.get(level, {}).get("sum", 0),
                })

    return results


def write_results(results, out_path):
    if not results:
        return
    keys = list(results[0].keys())
    with open(out_path, "w") as fh:
        fh.write("\t".join(keys) + "\n")
        for row in results:
            fh.write("\t".join("" if row[k] is None else str(row[k]) for k in keys) + "\n")


def main():
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("input", help="JSON file from prepare_cluster_data.py")
    p.add_argument("--output", default="cluster_analysis_results.tsv")
    args = p.parse_args()

    sys.stderr.write(f"Loading {args.input} ...\n")
    t_load = time.time()
    with open(args.input) as fh:
        data = json.load(fh)
    sys.stderr.write(f"Loaded in {time.time()-t_load:.2f}s\n")

    n_clusters = len(data["clusters"])
    n_attrs = len(data["alo_collection"])
    sys.stderr.write(f"  {n_clusters} clusters, {n_attrs} attributes\n")

    # --- Timed section ---
    t0 = time.time()
    results = run_analysis(data)
    elapsed = time.time() - t0
    # --- End timed section ---

    sys.stderr.write(f"Writing {args.output} ...\n")
    write_results(results, args.output)

    print(f"cluster_analysis_elapsed_s\t{elapsed:.6f}")
    print(f"clusters\t{n_clusters}")
    print(f"result_rows\t{len(results)}")
    sys.stderr.write(f"Done. {elapsed:.3f}s\n")


if __name__ == "__main__":
    main()

File: kinfin_stats/src/bin/run_cluster_analysis.rs
//! Rust cluster analysis benchmark binary.
//!
//! Reads the JSON file produced by prepare_cluster_data.py and runs the full
//! cluster analysis loop — identical logic to the Python run_cluster_analysis.py —
//! then writes a TSV result file.
//!
//! Timing covers only the analysis loop, excluding JSON I/O.
//!
//! Usage:
//!   ./run_cluster_analysis cluster_data.json [--output results.tsv]

use serde::{Deserialize, Serialize};
use serde_json::Value;
use statrs::distribution::{ContinuousCDF, Normal};
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs;
use std::io::{BufWriter, Write};
use std::time::Instant;

// ---------------------------------------------------------------------------
// JSON input structs
// ---------------------------------------------------------------------------

#[derive(Deserialize, Debug)]
struct Params {
    test: String,
    min_proteomes: usize,
    fuzzy_count: usize,
    fuzzy_fraction: f64,
    fuzzy_range: Vec<usize>,
    fastas_parsed: bool,
}

#[derive(Deserialize, Debug)]
struct AloLevel {
    proteomes_list: Vec<String>,
    proteome_count: usize,
}

#[derive(Deserialize, Debug)]
struct ClusterInput {
    cluster_id: String,
    singleton: bool,
    // proteome_id -> list of protein_ids
    protein_ids_by_proteome_id: HashMap<String, Vec<String>>,
}

#[derive(Deserialize, Debug)]
struct InputData {
    params: Params,
    // attribute -> level -> AloLevel (or null)
    alo_collection: HashMap<String, HashMap<String, Option<AloLevel>>>,
    clusters: Vec<ClusterInput>,
    protein_lengths: HashMap<String, u32>,
}

// ---------------------------------------------------------------------------
// Output row
// ---------------------------------------------------------------------------

struct ResultRow {
    cluster_id: String,
    attribute: String,
    level: String,
    cluster_type: &'static str,
    alo_status: &'static str,
    proteome_coverage: f64,
    cardinality: Option<&'static str>,
    mwu_pvalue: Option<f64>,
    mwu_log2_mean: Option<f64>,
    mean_alo: Option<f64>,
    mean_non_alo: Option<f64>,
    protein_count: usize,
    length_sum: u32,
}

// ---------------------------------------------------------------------------
// Mann-Whitney U (matches lib.rs implementation exactly)
// ---------------------------------------------------------------------------

fn normal_cdf(x: f64) -> f64 {
    Normal::new(0.0, 1.0).unwrap().cdf(x)
}

/// Exact two-sided p-value for Mann-Whitney U via DP over rank sequences.
/// Matches scipy.stats.mannwhitneyu method='exact' (no ties assumed).
/// m = min(n1, n2), n = max(n1, n2), u_obs_min = min(U1, U2).
fn mannwhitneyu_exact_pvalue(u_obs_min: usize, m: usize, n: usize) -> f64 {
    let max_u = m * n;
    let width = max_u + 1;
    let mut prev = vec![0.0f64; (m + 1) * width];
    let mut cur = vec![0.0f64; (m + 1) * width];
    for i in 0..=m {
        prev[i * width] = 1.0;
    }
    for _j in 1..=n {
        for v in cur.iter_mut() {
            *v = 0.0;
        }
        cur[0] = 1.0;
        for i in 1..=m {
            let base = i * width;
            let base_im1 = (i - 1) * width;
            for k in 0..=max_u {
                let from_x = cur[base_im1 + k];
                let from_y = if k >= i { prev[base + k - i] } else { 0.0 };
                cur[base + k] = from_x + from_y;
            }
        }
        std::mem::swap(&mut prev, &mut cur);
    }
    let base_m = m * width;
    let total: f64 = prev[base_m..base_m + width].iter().sum();
    if total == 0.0 {
        return 1.0;
    }
    let p_le: f64 = prev[base_m..base_m + u_obs_min + 1].iter().sum::<f64>() / total;
    (2.0 * p_le).min(1.0)
}

fn mannwhitneyu(x: &[f64], y: &[f64]) -> f64 {
    let n1 = x.len();
    let n2 = y.len();
    if n1 == 0 || n2 == 0 {
        return 1.0;
    }
    let mut combined: Vec<(f64, usize)> = Vec::with_capacity(n1 + n2);
    for &v in x {
        combined.push((v, 0));
    }
    for &v in y {
        combined.push((v, 1));
    }
    combined.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(std::cmp::Ordering::Equal));

    let mut ranks = vec![0.0_f64; combined.len()];
    let mut i = 0;
    while i < combined.len() {
        let mut j = i;
        while j < combined.len() && (combined[j].0 - combined[i].0).abs() < 1e-15 {
            j += 1;
        }
        let avg_rank = ((i + j) as f64 + 1.0) / 2.0;
        for idx in i..j {
            ranks[idx] = avg_rank;
        }
        i = j;
    }

    let mut rank_sum_x = 0.0_f64;
    for (idx, (_, group)) in combined.iter().enumerate() {
        if *group == 0 {
            rank_sum_x += ranks[idx];
        }
    }

    let n1f = n1 as f64;
    let n2f = n2 as f64;
    let u1 = rank_sum_x - n1f * (n1f + 1.0) / 2.0;
    let u2 = n1f * n2f - u1;
    let u_min = u1.min(u2);

    // Check for ties
    let has_ties = combined.windows(2).any(|w| (w[0].0 - w[1].0).abs() < 1e-15);

    // Exact method when min(n1, n2) <= 8 and no ties (matches scipy 'auto')
    if !has_ties && (n1 <= 8 || n2 <= 8) {
        let m = n1.min(n2);
        let n = n1.max(n2);
        return mannwhitneyu_exact_pvalue(u_min as usize, m, n);
    }

    // Asymptotic path with tie-corrected variance (matches scipy _get_mwu_z)
    let n_total = (n1 + n2) as f64;
    let mean_u = n1f * n2f / 2.0;

    // Tie correction: sum of (t^3 - t) over each tied run, where t = run length
    let tie_term: f64 = {
        let mut sum = 0.0f64;
        let mut i = 0;
        while i < combined.len() {
            let mut j = i + 1;
            while j < combined.len() && (combined[j].0 - combined[i].0).abs() < 1e-15 {
                j += 1;
            }
            let t = (j - i) as f64;
            sum += t * t * t - t;
            i = j;
        }
        sum
    };
    let var_u = n1f * n2f / 12.0 * ((n_total + 1.0) - tie_term / (n_total * (n_total - 1.0)));
    if var_u <= 0.0 {
        return 1.0;
    }
    // scipy: z = (U_max - 0.5 - mean_u) / s, p = 2*sf(z).
    // Equivalent with u_min: z = (u_min + 0.5 - mean_u) / s, p = 2*CDF(z) [z <= 0 typically].
    let z = (u_min + 0.5 - mean_u) / var_u.sqrt();
    (2.0 * normal_cdf(z)).clamp(0.0, 1.0)
}

// ---------------------------------------------------------------------------
// Stat dispatch
// ---------------------------------------------------------------------------

fn statistic(
    count1: &[i32],
    count2: &[i32],
    test: &str,
    min_proteomes: usize,
) -> (Option<f64>, Option<f64>, Option<f64>, Option<f64>) {
    let c1: Vec<f64> = count1
        .iter()
        .filter(|&&c| c > 0)
        .map(|&c| c as f64)
        .collect();
    let c2: Vec<f64> = count2
        .iter()
        .filter(|&&c| c > 0)
        .map(|&c| c as f64)
        .collect();
    if c1.len() < min_proteomes || c2.len() < min_proteomes {
        return (None, None, None, None);
    }
    let mean_c1 = c1.iter().sum::<f64>() / c1.len() as f64;
    let mean_c2 = c2.iter().sum::<f64>() / c2.len() as f64;
    let log2_mean = (mean_c1 / mean_c2).log2();

    // Check if all values equal
    let s1: HashSet<u64> = c1.iter().map(|&x| x.to_bits()).collect();
    let s2: HashSet<u64> = c2.iter().map(|&x| x.to_bits()).collect();
    let pvalue = if s1.len() == 1 && s2.len() == 1 && s1 == s2 {
        1.0
    } else {
        match test {
            "mannwhitneyu" => mannwhitneyu(&c1, &c2),
            // other tests fall back to MWU for now (benchmark focus is mannwhitneyu)
            _ => mannwhitneyu(&c1, &c2),
        }
    };

    (Some(pvalue), Some(log2_mean), Some(mean_c1), Some(mean_c2))
}

// ---------------------------------------------------------------------------
// Cardinality
// ---------------------------------------------------------------------------

fn get_cardinality(
    alo_counts: &[i32],
    fuzzy_count: usize,
    fuzzy_fraction: f64,
    fuzzy_range: &HashSet<usize>,
) -> Option<&'static str> {
    let non_zero: Vec<i32> = alo_counts.iter().copied().filter(|&c| c > 0).collect();
    if non_zero.is_empty() {
        return None;
    }
    if non_zero.iter().all(|&c| c == 1) {
        return Some("true");
    }
    let fc = fuzzy_count as i32;
    if non_zero.contains(&fc)
        || (non_zero.len() as f64 >= alo_counts.len() as f64 * fuzzy_fraction
            && non_zero
                .iter()
                .all(|c| fuzzy_range.contains(&(*c as usize))))
    {
        return Some("fuzzy");
    }
    None
}

// ---------------------------------------------------------------------------
// Length stats
// ---------------------------------------------------------------------------

fn get_length_stats(
    protein_ids: &[String],
    protein_lengths: &HashMap<String, u32>,
    fastas_parsed: bool,
) -> (u32, f64, f64, f64) {
    if !fastas_parsed || protein_ids.is_empty() {
        return (0, 0.0, 0.0, 0.0);
    }
    let lengths: Vec<u32> = protein_ids
        .iter()
        .filter_map(|id| protein_lengths.get(id).copied())
        .filter(|&l| l > 0)
        .collect();
    if lengths.is_empty() {
        return (0, 0.0, 0.0, 0.0);
    }
    let sum: u32 = lengths.iter().sum();
    let mean = sum as f64 / lengths.len() as f64;
    let mut sorted = lengths.clone();
    sorted.sort_unstable();
    let n = sorted.len();
    let median = (sorted[n / 2] as f64 + sorted[(n - 1) / 2] as f64) / 2.0;
    let sd = if lengths.len() > 1 {
        (lengths
            .iter()
            .map(|&x| (x as f64 - mean).powi(2))
            .sum::<f64>()
            / lengths.len() as f64)
            .sqrt()
    } else {
        0.0
    };
    (sum, mean, median, sd)
}

// ---------------------------------------------------------------------------
// Main analysis loop
// ---------------------------------------------------------------------------

fn run_analysis(data: &InputData) -> Vec<ResultRow> {
    let params = &data.params;
    let test = params.test.as_str();
    let min_proteomes = params.min_proteomes;
    let fuzzy_count = params.fuzzy_count;
    let fuzzy_fraction = params.fuzzy_fraction;
    let fuzzy_range: HashSet<usize> = params.fuzzy_range.iter().copied().collect();
    let fastas_parsed = params.fastas_parsed;
    let attributes: Vec<&String> = data.alo_collection.keys().collect();

    let mut results = Vec::new();

    for cluster in &data.clusters {
        let cluster_id = &cluster.cluster_id;
        let singleton = cluster.singleton;
        let pids_by_proteome = &cluster.protein_ids_by_proteome_id;

        for attribute in &attributes {
            let levels = &data.alo_collection[*attribute];

            // Pass 1: build per-level data
            let mut protein_ids_by_level: HashMap<&str, Vec<&str>> = HashMap::new();
            let mut length_stats_by_level: HashMap<&str, (u32, f64, f64, f64)> = HashMap::new();
            let mut explicit_count_by_level: HashMap<&str, HashMap<&str, i32>> = HashMap::new();
            let mut implicit_by_level: HashMap<&str, bool> = HashMap::new(); // level -> has any

            for (level, alo_opt) in levels {
                let alo = match alo_opt {
                    Some(a) => a,
                    None => continue,
                };
                let mut pids_for_level: Vec<&str> = Vec::new();
                let mut count_by_proteome: HashMap<&str, i32> = HashMap::new();
                let mut has_implicit = false;

                for proteome_id in &alo.proteomes_list {
                    let pids = pids_by_proteome.get(proteome_id);
                    let count = pids.map_or(0, |v| v.len()) as i32;
                    count_by_proteome.insert(proteome_id.as_str(), count);
                    if let Some(pids) = pids {
                        pids_for_level.extend(pids.iter().map(String::as_str));
                        if count > 0 {
                            has_implicit = true;
                        }
                    }
                }

                let pids_owned: Vec<String> =
                    pids_for_level.iter().map(|s| s.to_string()).collect();
                let ls = get_length_stats(&pids_owned, &data.protein_lengths, fastas_parsed);
                length_stats_by_level.insert(level.as_str(), ls);
                protein_ids_by_level.insert(level.as_str(), pids_for_level);
                explicit_count_by_level.insert(level.as_str(), count_by_proteome);
                if has_implicit {
                    implicit_by_level.insert(level.as_str(), true);
                }
            }

            let cluster_type: &'static str = if singleton {
                "singleton"
            } else if implicit_by_level.len() > 1 {
                "shared"
            } else {
                "specific"
            };

            // Pass 2: per-level stats
            for (level, alo_opt) in levels {
                let alo = match alo_opt {
                    Some(a) => a,
                    None => continue,
                };
                let proteome_count = alo.proteome_count;
                let alo_status: &'static str = if implicit_by_level.contains_key(level.as_str()) {
                    "present"
                } else {
                    "absent"
                };
                let implicit_count = if alo_status == "present" {
                    explicit_count_by_level[level.as_str()]
                        .values()
                        .filter(|&&c| c > 0)
                        .count()
                } else {
                    0
                };
                let proteome_coverage = implicit_count as f64 / proteome_count as f64;

                let mut cardinality: Option<&'static str> = None;
                let mut mwu_pvalue: Option<f64> = None;
                let mut mwu_log2_mean: Option<f64> = None;
                let mut mean_alo: Option<f64> = None;
                let mut mean_non_alo: Option<f64> = None;

                if alo_status == "present" && cluster_type != "singleton" {
                    let alo_counts: Vec<i32> = explicit_count_by_level[level.as_str()]
                        .values()
                        .copied()
                        .collect();
                    cardinality =
                        get_cardinality(&alo_counts, fuzzy_count, fuzzy_fraction, &fuzzy_range);

                    if cluster_type == "shared" {
                        let non_alo_counts: Vec<i32> = explicit_count_by_level
                            .iter()
                            .filter(|(&l, _)| l != level.as_str())
                            .flat_map(|(_, counts)| counts.values().copied())
                            .collect();
                        let (pv, lm, ma, mn) =
                            statistic(&alo_counts, &non_alo_counts, test, min_proteomes);
                        mwu_pvalue = pv;
                        mwu_log2_mean = lm;
                        mean_alo = ma;
                        mean_non_alo = mn;
                    }
                }

                let pids = protein_ids_by_level
                    .get(level.as_str())
                    .map_or(0, |v| v.len());
                let ls = length_stats_by_level
                    .get(level.as_str())
                    .copied()
                    .unwrap_or((0, 0.0, 0.0, 0.0));

                results.push(ResultRow {
                    cluster_id: cluster_id.clone(),
                    attribute: attribute.to_string(),
                    level: level.clone(),
                    cluster_type,
                    alo_status,
                    proteome_coverage,
                    cardinality,
                    mwu_pvalue,
                    mwu_log2_mean,
                    mean_alo,
                    mean_non_alo,
                    protein_count: pids,
                    length_sum: ls.0,
                });
            }
        }
    }

    results
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: run_cluster_analysis <cluster_data.json> [--output <out.tsv>]");
        std::process::exit(1);
    }

    let input_path = &args[1];
    let output_path = args
        .windows(2)
        .find(|w| w[0] == "--output")
        .map(|w| w[1].clone())
        .unwrap_or_else(|| "cluster_analysis_results_rust.tsv".to_string());

    eprintln!("Loading {} ...", input_path);
    let t_load = Instant::now();
    let raw = fs::read_to_string(input_path).expect("Cannot read input file");
    let data: InputData = serde_json::from_str(&raw).expect("JSON parse error");
    eprintln!("Loaded in {:.2}s", t_load.elapsed().as_secs_f64());

    let n_clusters = data.clusters.len();
    let n_attrs = data.alo_collection.len();
    eprintln!("  {} clusters, {} attributes", n_clusters, n_attrs);

    // --- Timed section ---
    let t0 = Instant::now();
    let results = run_analysis(&data);
    let elapsed = t0.elapsed().as_secs_f64();
    // --- End timed section ---

    eprintln!("Writing {} ...", output_path);
    let f = fs::File::create(&output_path).expect("Cannot create output file");
    let mut w = BufWriter::new(f);
    writeln!(w, "cluster_id\tattribute\tlevel\tcluster_type\talo_status\tproteome_coverage\tcardinality\tmwu_pvalue\tmwu_log2_mean\tmean_alo\tmean_non_alo\tprotein_count\tlength_sum").unwrap();
    for row in &results {
        writeln!(
            w,
            "{}\t{}\t{}\t{}\t{}\t{:.6}\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
            row.cluster_id,
            row.attribute,
            row.level,
            row.cluster_type,
            row.alo_status,
            row.proteome_coverage,
            row.cardinality.unwrap_or(""),
            row.mwu_pvalue
                .map_or(String::new(), |v| format!("{:.8}", v)),
            row.mwu_log2_mean
                .map_or(String::new(), |v| format!("{:.8}", v)),
            row.mean_alo.map_or(String::new(), |v| format!("{:.4}", v)),
            row.mean_non_alo
                .map_or(String::new(), |v| format!("{:.4}", v)),
            row.protein_count,
            row.length_sum,
        )
        .unwrap();
    }

    println!("cluster_analysis_elapsed_s\t{:.6}", elapsed);
    println!("clusters\t{}", n_clusters);
    println!("result_rows\t{}", results.len());
    eprintln!("Done. {:.3}s", elapsed);
}
